custom-routes-vis/src/main/webapp/script.js

```diff
@@ -143,9 +156,140 @@ function hideMarker(markerName) {
   document.getElementById('generate-routes').style.display = 'none';
 }
 
+/** Removes all routes from the DOM. */
+function clearRoutes() {
+  for (let i = 0; i < displayedRoutes.length; i++) {
+    displayedRoutes[i].setMap(null);
+  }
+  displayedRoutes = [];
+  selectedRouteNum = 0;
+}
+
 /**
- * Generates routes from origin to destination by calling Directions API.
+ * Generates routes from origin to destination by calling the selected API.
+ * Precondition: two markers exist in the originDestinationMarkers array.
  */
 function generateRoutes() {
-  // TODO(chenyuz): Write this code in separate branch.
-}
\ No newline at end of file
+  clearRoutes();
+
+  // Get coordinates of origin and destination.
+  let startLatLng = originDestinationMarkers[0].getPosition();
+  let endLatLng = originDestinationMarkers[1].getPosition();
+  let origin = startLatLng.lat() + ',' + startLatLng.lng();
+  let destination = endLatLng.lat() + ',' + endLatLng.lng();
+
+  let serviceEndpoint = document.getElementById('service-endpoint').value;
+  let apiKey = document.getElementById('api-key').value;
+
+  // Get the routes and display them on map.
+  fetch('/get-directions?origin=' + origin + '&destination=' + destination
+      + '&endpoint=' + serviceEndpoint + '&apiKey=' + apiKey)
+      .then(response => response.json()).then(directions => {
+        // Log the response status from the Directions API.
+        // TODO(chenyuz): find equivalence for Routes Preferred API.
+        console.log(directions.status);
+
+        let routes = directions.routes;
+        console.log('num routes:', routes.length);
+
+        for (let routeNum = 0; routeNum < routes.length; routeNum++) {
+          createRoutePolyline(routeNum, routes[routeNum], 
+              serviceEndpoint === 'directions');
+        }
+      });
+}
+
+/**
+ * Creates a polyline for a route on the map.
+ * @param {num} routeNum The index of the route in the order as it is returned
+ * from the API.
+ * @param {!JSON} routeJson JSON object containing all information of the target
+ * route.
+ * @param {bool} directionsApi Whether the route is obtained by the Directions
+ * API; if false, assume it is obtained by the Routes Preferred API.
+ */
+function createRoutePolyline(routeNum, routeJson, directionsApi) {
+  let routeLegs = routeJson.legs;
+  let routeCoordinates = [];
+
+  // Total duration of route in seconds.
+  let totalDuration = directionsApi ? 0 : routeJson.duration;
+  // Total distance of route in meters.
+  let totalDistance = directionsApi ? 0 : routeJson.distanceMeters;
+  // Note: Routes Preferred has duration and distanceMeters attributes for each
+  // route, but Directions only has those for each leg of the route.
+
+  for (let i = 0; i < routeLegs.length; i++) {
+    let legSteps = routeLegs[i].steps;
+    console.log('num steps', legSteps.length);
+
+    for (let j = 0; j < legSteps.length; j++) {
```

Comment:
could you use forEach?

Reply:
If you're replacing loops it's best to use for-of loops, 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of
'With ES6, the language now has three different kinds of for loops. All may be used, though for-of loops should be preferred when possible.' 
https://google.github.io/styleguide/jsguide.html#formatting-line-wrapping